book/src/types/rust_types.md

```diff
@@ -0,0 +1,246 @@
+goes through the variants in a bit more detail, and in particular
+describes why each variant exists.
+
+### Application types
+
+The `Apply` variant contains an `ApplicationTy`. These are kind of the
+"normal Rust types", like `Vec<u32>` or `f32`. They consist of a "type
+name" (in our examples, `Vec` and `f32` respecively) and zero or more
+generic arguments (respectively, `[u32]` and `[]`).
+
+They are equal to other types (modulo aliases, see below) iff they
+have the same "type name" and the generic arguments are
+recursively equal
+
+### Placeholders
+
+The `Placeholder` variant contains a `PlaceholderIndex` type. It
+represents a generic type that is being treated abstractly or -- more
+generally -- the result of a "type function" that cannot be
+evaluated. For example, when typing the body of a generic function
+like `fn foo<T: Iterator>`, the type `T` would be represented with a
+placeholder. Similarly, in that same function, the associated type
+`T::Item` might be represented with a placeholder.
+    
+Like application types, placeholder *types* are only known to be
+equal.
+
+However, we choose not to represent placeholder types as type names
+because they need to be created during type unification and other
+operations, and hence that would require treating `TypeName` less opaquely.
+
+Moreover, when proving negative goals, e.g., `not { Implemented(T:
+Trait) }`, placeholders are treated quite differently from application
+types, since they do not (in fact) represent a known type. When
+solving negative goals, placeholderes are replaced with inference
+variables -- the idea is that this goal is only true if there is *no
+type* `T` that implements `Trait`. Therefore, if we can find no
+answeres for `exists<T> { Implemented(T: Trait) }`, then we know that
+the negation is true. (Note that this means that e.g. `forall<X> { X =
+i32 }` is false but so is `forall<X> { not { X = i32 } }`.)
+
+### Inference variables
+
+The `InferenceVar` variant wraps an `InferenceVar` type.  This
+represents a type whose value is being inferred. The value of an
+inference variables may be "known" or "not known", but that state is
+stored externally, in the inference context (see the section on
+inference below).
+
+When equating, inference variables are treated specially in that they
+become bound (or, if they have already been bound, they are replaced
+with their value).
+
+Inference variables are also integral to canonicalization and
+other types.
+
+### Dyn types
+
+The `Dyn` variant wraps a `DynTy` and represents a `dyn Trait`
+type. In chalk, these are represented as an existential type where we
+store the predicates that are known to be true. So a type like `dyn
+Write` would be represented as, effectively, an `exists<T> { T: Write
+}` type.
+  
+When equating, two `dyn P` and `dyn Q` types are equal if `P = Q` --
+i.e., they have the same bounds. Note that -- for this purpose --
+ordering of bounds is significant, and chalk assumes that a canonical
```

Comment:
What bounds? Like `T` and `U` in `forall<T, U>`? Why are these significant? (Or, I guess, why aren't they significant elsewhere?)

Reply:
I'll try to clarify -- the word "bounds" here is referring to the traits referenced of the dyn type. I tend to use bound probably for many purposes :P but one of them is for a "trait that a type must implement", or a "where clause without the self type" (i.e., what comes after a `:` in Rust syntax). It's basically "something that constraints (bounds) the choice of a type parameter":

```
where
    P: PartialEq<u32>
    // -------------- bound is just `PartialEq<u32>`
 // ----------------- where clause includes the `P:`
```

So e.g. in `dyn Debug`, the "bound" is the `Debug` trait

Reply:
It's unclear to me how the ordering of bounds here (maybe we should think of a better distinction vs. binders?) is significant or *why*. We should maybe discuss this?

Reply:
I've at least clarified the current *behavior* -- I don't know that it's correct =)